### PR TITLE
pre render glyphs

### PR DESCRIPTION
**Pull request by tronical**

### CODE REVIEW

File: api/rs/slint/lib.rs [hunk 1]
```diff
@@ -268,6 +268,14 @@ pub fn register_font_from_path<P: AsRef<std::path::Path>>(
     i_slint_backend_selector::backend().register_font_from_path(path.as_ref())
 }
 
+/// This function can be used to register a pre-rendered, embedded bitmap font with Slint,
+/// for use with the `font-family` property.
+#[doc(hidden)]
+#[cfg(not(feature = "std"))]
```
[the remark this answers is not in the document]
Comment (ogoffart) [reply api/rs/slint/lib.rs:274]: yes

File: internal/backends/mcu/fonts.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    let font = FONTS.with(|fonts| {
+        let fonts = fonts.borrow();
+        let fallback_font =
+            *fonts.first().expect("internal error: cannot render text without fonts");
+
+        request.family.as_ref().map_or(fallback_font, |requested_family| {
+            fonts
+                .iter()
+                .find(|bitmap_font| {
+                    core::str::from_utf8(bitmap_font.family_name.as_slice()).unwrap()
```
Comment (ogoffart) [inline internal/backends/mcu/fonts.rs:29]: I think, long term, we should not do the font lookup of familly at runtime and have the property on text already be a resolved font index or something.

But let's keep it as is for now.

File: internal/backends/mcu/fonts.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+use core::cell::RefCell;
+use core::convert::TryFrom;
+
+#[cfg(all(not(feature = "std"), feature = "unsafe_single_core"))]
+use i_slint_core::thread_local_ as thread_local;
+
+use i_slint_core::graphics::{BitmapFont, BitmapGlyph, BitmapGlyphs, FontRequest, Size};
+
+thread_local! {
+    static FONTS: RefCell<Vec<&'static BitmapFont>> = RefCell::default()
```
[the remark this answers is not in the document]
Comment (tronical) [reply internal/backends/mcu/fonts.rs:14]: Good idea

File: internal/backends/mcu/fonts.rs (hunk 2 above)
Comment (tronical) [reply internal/backends/mcu/fonts.rs:29]: Yes, we should try to resolve at compile time what we can.

File: internal/backends/mcu/fonts.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                        == requested_family.as_str()
+                })
+                .unwrap_or(&fallback_font)
+        })
+    });
+
+    let requested_pixel_size = request
+        .pixel_size
+        .and_then(|size| u16::try_from(size as i64).ok())
+        .unwrap_or(DEFAULT_FONT_SIZE);
```
[the remark this answers is not in the document]
Comment (tronical) [reply internal/backends/mcu/fonts.rs:39]: Indeed

**Review by ogoffart — APPROVED** [review]
(The commits can be put in master)

**Comment by tronical** [comment]
Merged into master as commit bd2a62b0c2d7a4de45d77ad9994eed5611fedf43 and `wip/mcu` rebased.

**Pull request closed without merging**